### PR TITLE
[CBRD-20379] New cleanup criteria for XASL cache

### PR DESCRIPTION
**Pull request by funkydvd**
http://jira.cubrid.org/browse/CBRD-20379
XASL Cache was cleaned only when the maximum capacity is reached. I introduced another criteria, based on time. After a certain amount of time, the oldest entries are cleaned. I introduced a new system parameter for the time threshold and made a new cleanup type. For this type, it is used an array for storing the entries instead of a binary heap.

### CODE REVIEW

File: src/query/xasl_cache.c [hunk 1]
```diff
@@ -68,51 +68,70 @@ struct xcache_stats
...
   0,				/* cleanup_flag */
   NULL,				/* cleanup_bh */
+  NULL,				/* cleanup_array */
   XCACHE_STATS_INITIALIZER
 };
 
 /* Create macro's for xcache_Global fields to access them as if they were global variables. */
 #define xcache_Enabled xcache_Global.enabled
 #define xcache_Soft_capacity xcache_Global.soft_capacity
+#define xcache_time_threshold xcache_Global.time_threshold
```
Comment (eseokoh) [inline src/query/xasl_cache.c]: Please rename it as `xcache_Time_threshold`.

File: src/query/xasl_cache.c [hunk 2]
```diff
@@ -68,51 +68,70 @@ struct xcache_stats
...
+#define xcache_last_cleaned_time xcache_Global.last_cleaned_time
```
Comment (eseokoh) [inline src/query/xasl_cache.c]: Please rename it as `xcache_Last_cleaned_time`.

File: src/query/xasl_cache.c [hunk 3]
```diff
@@ -68,51 +68,70 @@ struct xcache_stats
...
 #define xcache_Ht xcache_Global.ht
 #define xcache_Ht_freelist xcache_Global.freelist
 #define xcache_Entry_count xcache_Global.entry_count
 #define xcache_Log xcache_Global.logging_enabled
 #define xcache_Max_clones xcache_Global.max_clones
 #define xcache_Cleanup_flag xcache_Global.cleanup_flag
 #define xcache_Cleanup_bh xcache_Global.cleanup_bh
+#define xcache_cleanup_array xcache_Global.cleanup_array
```
Comment (eseokoh) [inline src/query/xasl_cache.c]: Please rename it as `xcache_Cleanup_array`.

File: src/query/xasl_cache.c [hunk 4]
```diff
@@ -276,6 +291,8 @@ xcache_initialize (THREAD_ENTRY * thread_p)
   xcache_Cleanup_flag = 0;
   xcache_Cleanup_bh = bh_create (thread_p, XCACHE_CLEANUP_NUM_ENTRIES (xcache_Soft_capacity),
 				 sizeof (XCACHE_CLEANUP_CANDIDATE), xcache_compare_cleanup_candidates, NULL);
+  xcache_cleanup_array = malloc (xcache_Soft_capacity * sizeof (XCACHE_CLEANUP_CANDIDATE));
```
Comment (eseokoh) [inline src/query/xasl_cache.c]: Please move it to next to `  if (xcache_Cleanup_bh == NULL)`.
The block to allocate `xcache_Cleanup_bh` is very special to use the global private heap.

Please also add error handling of `malloc`.

File: src/query/xasl_cache.c [hunk 5]
```diff
@@ -1443,7 +1459,10 @@ xcache_insert (THREAD_ENTRY * thread_p, const COMPILE_CONTEXT * context, XASL_ST
     }
   else
     {
-      if (xcache_Soft_capacity < xcache_Entry_count && xcache_Cleanup_flag == 0)
+      gettimeofday (&current_time, NULL);
+      if ((xcache_Soft_capacity < xcache_Entry_count
+	   || TIME_DIFF_SEC (current_time, xcache_last_cleaned_time) > xcache_time_threshold)
```
Comment (eseokoh) [inline src/query/xasl_cache.c]: I'd like to propose to introduce a function something like this.

```c
typedef enum xcache_cleanup_reason XCACHE_CLEANUP_REASON;
enum xcache_cleanup_reason
{
  XCACHE_CLEANUP_NONE; /* no cleanup is required */
  XCACHE_CLEANUP_FULL;
  XCACHE_CLEANUP_TIMEOUT;
};

static XCACHE_CLEANUP_REASON
xcache_need_cleanup (void)
{
  if (xcache_Soft_capacity < xcache_Entry_count)
    {
      return XCACHE_CLEANUP_FULL;
    }
  else
    {
      ...
      if (TIME_DIFF_SET (current_time, xcache_last_cleaned_time) > xcache_time_threshold)
        {
          return XCACHE_CLEANUP_TIMEOUT;
        }
    }

  return XCACHE_CLEANUP_NONE;
}
```

You may think of better names for enums and functions :)

Then, you can rewrite `if` with the new function. 
```c
  if (xcache_need_cleanup () != XCACHE_CLEANUP_NONE
      && xcache_Cleanup_flag == 0)
```

File: src/query/xasl_cache.c [hunk 6]
```diff
@@ -1860,7 +1881,10 @@ xcache_cleanup (THREAD_ENTRY * thread_p)
       /* Somebody else does the cleanup. */
       return;
     }
-  if (xcache_Entry_count <= xcache_Soft_capacity)
+
+  gettimeofday (&current_time, NULL);
+  if (xcache_Entry_count <= xcache_Soft_capacity
+      && TIME_DIFF_SEC (current_time, xcache_last_cleaned_time) <= xcache_time_threshold)
```
Comment (eseokoh) [inline src/query/xasl_cache.c]: You can rewrite it as follows:
```c
  need_cleanup = xcache_need_cleanup ();
  if (need_cleanup == XCACHE_CLEANUP_NONE)
```

File: src/query/xasl_cache.c [hunk 7]
```diff
@@ -1874,82 +1898,121 @@ xcache_cleanup (THREAD_ENTRY * thread_p)
   perfmon_inc_stat (thread_p, PSTAT_PC_NUM_FULL);
 
   /* How many entries do we need to cleanup? */
-  cleanup_count = (int) (XCACHE_CLEANUP_RATIO * xcache_Soft_capacity) + (xcache_Entry_count - xcache_Soft_capacity);
-  if (cleanup_count <= 0)
+  if (xcache_Entry_count > xcache_Soft_capacity)
```
Comment (eseokoh) [inline src/query/xasl_cache.c]:
```c
  if (need_cleanup == XCACHE_CLEANUP_FULL)
```

File: src/query/xasl_cache.c [hunk 8]
```diff
@@ -1874,82 +1898,121 @@ xcache_cleanup (THREAD_ENTRY * thread_p)
   perfmon_inc_stat (thread_p, PSTAT_PC_NUM_FULL);
```
Comment (eseokoh) [inline src/query/xasl_cache.c]: Please move it to block of `XCACHE_CLEANUP_FULL`.

File: src/query/xasl_cache.c [hunk 9]
```diff
@@ -1874,82 +1898,121 @@ xcache_cleanup (THREAD_ENTRY * thread_p)
...
+  else
     {
-      candidate.xid = xcache_entry->xasl_id;
-      candidate.time_last_used = xcache_entry->time_last_used;
+      count = 0;
+      /* Collect candidates for cleanup. */
+      lf_hash_create_iterator (&iter, t_entry, &xcache_Ht);
 
-      if (candidate.xid.cache_flag & XCACHE_ENTRY_FLAGS_MASK)
+      while ((xcache_entry = (XASL_CACHE_ENTRY *) lf_hash_iterate (&iter)) != NULL)
```
Comment (eseokoh) [inline src/query/xasl_cache.c]: `xcache_Cleanup_array` may be overrun. 

```c
while ((xcache_entry = (XASL_CACHE_ENTRY *) lf_hash_iterate (&iter)) != NULL
       && count < xcache_Soft_capacity)
```

File: src/query/xasl_cache.c [hunk 10]
```diff
@@ -1443,7 +1487,8 @@ xcache_insert (THREAD_ENTRY * thread_p, const COMPILE_CONTEXT * context, XASL_ST
     }
   else
     {
-      if (xcache_Soft_capacity < xcache_Entry_count && xcache_Cleanup_flag == 0)
+      gettimeofday (&current_time, NULL);
```
Comment (eseokoh) [inline src/query/xasl_cache.c]: You don't need it :)
Please remove `current_time`.

File: src/query/xasl_cache.c [hunk 11]
```diff
@@ -1998,6 +1997,7 @@ xcache_cleanup (THREAD_ENTRY * thread_p)
       count = 0;
       /* Collect candidates for cleanup. */
       lf_hash_create_iterator (&iter, t_entry, &xcache_Ht);
+      gettimeofday (&current_time, NULL);
```
Comment (eseokoh) [inline src/query/xasl_cache.c]: please also remove the line :)

Comment (funkydvd) [inline src/query/xasl_cache.c]: I use this variable for finding the difference between the current time and last time used (line 2008), I think it is useful :)

**Comment by eseokoh** [comment]
You can merge the change if Remzi has no other feedback.
Thank you :)

**Pull request merged**